File: rs/src/getx.rs
// getx.rs — rich-path lookup, ported from go/getx.go / ts/src/util/basic.ts

use regex::Regex;

/// Rich-path lookup. Supports dot/space-separated navigation, ancestry (`:`),
/// comparison filters (`=`, `!=`, `<`, `<=`, `>`, `>=`, `==`, `~`),
/// array filters (`?`), and quoted segments.
///
/// Returns `serde_json::Value::Null` for any miss or invalid path.
pub fn getx(root: &serde_json::Value, path: &serde_json::Value) -> serde_json::Value {
    if root.is_null() {
        return serde_json::Value::Null;
    }
    if !matches!(root, serde_json::Value::Object(_) | serde_json::Value::Array(_)) {
        return serde_json::Value::Null;
    }

    let tokens: Vec<String> = match path {
        serde_json::Value::Null => return serde_json::Value::Null,
        serde_json::Value::String(s) => getx_tokenize(s),
        serde_json::Value::Array(arr) => arr.iter().map(|v| crate::util::spec_sprint(v)).collect(),
        _ => return serde_json::Value::Null,
    };

    if tokens.is_empty() {
        return serde_json::Value::Null;
    }

    let mut node: serde_json::Value = root.clone();
    let mut out: serde_json::Value = serde_json::Value::Null;
    let mut ancestry = false;
    let mut i = 0;

    while i < tokens.len() {
        if node.is_null() { break; }

        let t0 = &tokens[i];
        let t1 = tokens.get(i + 1).map(|s| s.as_str()).unwrap_or("");

        // Compare op: t1 is an operator
        if !t1.is_empty() && is_compare_op(t1) {
            let val = getx_index(&node, t0);
            let arg_raw = tokens.get(i + 2).map(|s| s.as_str()).unwrap_or("");
            let pass = getx_compare(&val, t1, arg_raw);
            if pass {
                i += 2;
            } else {
                node = serde_json::Value::Null;
            }
            if !(ancestry && !node.is_null()) {
                out = node.clone();
            }
            i += 1;
            continue;
        }

        // Ancestry op: t1 is ":"
        if t1 == ":" {
            let next = tokens.get(i + 2).map(|s| s.as_str()).unwrap_or("");
            if next != "=" {
                if !ancestry {
                    out = node.clone();
                }
                node = getx_index(&node, t0);
                if node.is_null() {
                    out = serde_json::Value::Null;
                }
            }
            ancestry = true;
            i += 2;
            continue;
        }

        // Array filter
        if t0 == "?" {
            let ftokens: Vec<String> = tokens[i + 1..].to_vec();
            // Find filter end: two adjacent identifier tokens
            let mut j = 0;
            while j < ftokens.len() {
                if j + 1 < ftokens.len()
                    && is_ident(&ftokens[j])
                    && is_ident(&ftokens[j + 1])
                {
                    j += 1;
                    break;
                }
                j += 1;
            }
            let ftokens = &ftokens[..j];

            let children = iter_children(&node);
            let filtered: Vec<_> = children
                .into_iter()
                .filter(|(_, v)| {
                    let path_val = serde_json::Value::Array(
                        ftokens.iter().map(|s| serde_json::Value::String(s.clone())).collect(),
                    );
                    !getx(v, &path_val).is_null()
                })
                .collect();

            node = rebuild(&node, &filtered);
            out = node.clone();
            i += j + 1;
            continue;
        }

        // Normal navigation
        if !t1.is_empty() {
            node = getx_index(&node, t0);
            if ancestry {
                ancestry = false;
                if node.is_null() {
                    out = serde_json::Value::Null;
                } else {
                    node = out.clone();
                }
            }
            i += 1;
            continue;
        }

        // Last token
        node = getx_index(&node, t0);
        if !(ancestry && !node.is_null()) {
            out = node.clone();
        }
        i += 1;
    }

    out
}

/// Token regex: mirrors `src/util/basic.ts:120`
fn getx_tokenize(p: &str) -> Vec<String> {
    // Mirrors Go getxTokenRE: `\s*("(\\.|[^"\\])*"|[\w\d_]+|\s+|[^\w\d_]+)\s*`
    let re = Regex::new(r#"\s*("(?:\\.|[^"\\])*"|[\w\d_]+|\s+|[^\w\d_]+)\s*"#).unwrap();
    let mut out = Vec::new();
    for cap in re.captures_iter(p) {
        let tok = &cap[1];
        if tok.is_empty() { continue; }
        if tok.trim().is_empty() || tok.trim_matches('.').is_empty() { continue; }
        // Strip surrounding double-quotes
        if tok.starts_with('"') && tok.ends_with('"') && tok.len() >= 2 {
            out.push(tok[1..tok.len() - 1].to_string());
        } else {
            out.push(tok.to_string());
        }
    }
    out
}

fn is_compare_op(t: &str) -> bool {
    matches!(t, "=" | "!=" | "<" | "<=" | ">" | ">=" | "==" | "~")
}

fn is_ident(t: &str) -> bool {
    !t.is_empty() && t.bytes().all(|b| b.is_ascii_alphanumeric() || b == b'_')
}

/// Index a JSON value by a string key (map) or integer index (array).
pub fn getx_index(node: &serde_json::Value, key: &str) -> serde_json::Value {
    match node {
        serde_json::Value::Object(m) => m.get(key).cloned().unwrap_or(serde_json::Value::Null),
        serde_json::Value::Array(a) => {
            if let Ok(i) = key.parse::<usize>() {
                a.get(i).cloned().unwrap_or(serde_json::Value::Null)
            } else {
                serde_json::Value::Null
            }
        }
        _ => serde_json::Value::Null,
    }
}

fn getx_compare(val: &serde_json::Value, op: &str, arg_raw: &str) -> bool {
    let arg: serde_json::Value = match arg_raw {
        "true" => serde_json::Value::Bool(true),
        "false" => serde_json::Value::Bool(false),
        s if s.starts_with('"') && s.ends_with('"') && s.len() >= 2 => {
            serde_json::Value::String(s[1..s.len() - 1].to_string())
        }
        s => serde_json::Value::String(s.to_string()),
    };

    let val_s = crate::util::spec_sprint(val);
    let arg_s = crate::util::spec_sprint(&arg);
    let vn: Option<f64> = val_s.parse().ok();
    let an: Option<f64> = arg_s.parse().ok();
    let both_num = vn.is_some() && an.is_some();

    let val_str = val.as_str();
    let arg_str = arg.as_str();
    let both_str = val_str.is_some() && arg_str.is_some();

    match op {
        "<" => {
            if both_str { val_str.unwrap() < arg_str.unwrap() }
            else { both_num && vn.unwrap() < an.unwrap() }
        }
        "<=" => {
            if both_str { val_str.unwrap() <= arg_str.unwrap() }
            else { both_num && vn.unwrap() <= an.unwrap() }
        }
        ">" => {
            if both_str { val_str.unwrap() > arg_str.unwrap() }
            else { both_num && vn.unwrap() > an.unwrap() }
        }
        ">=" => {
            if both_str { val_str.unwrap() >= arg_str.unwrap() }
            else { both_num && vn.unwrap() >= an.unwrap() }
        }
        "=" => {
            if both_num { vn.unwrap() == an.unwrap() }
            else { val_s == arg_s }
        }
        "==" => val == &arg,
        "!=" => {
            if both_num { vn.unwrap() != an.unwrap() }
            else { val_s != arg_s }
        }
        "~" => {
            Regex::new(&arg_s).map(|re| re.is_match(&val_s)).unwrap_or(false)
        }
        _ => false,
    }
}

/// Returns children as (key, value) pairs, sorted by key for maps.
fn iter_children(node: &serde_json::Value) -> Vec<(String, serde_json::Value)> {
    match node {
        serde_json::Value::Object(m) => {
            let mut keys: Vec<String> = m.keys().cloned().collect();
            keys.sort();
            keys.into_iter().map(|k| { let v = m[&k].clone(); (k, v) }).collect()
        }
        serde_json::Value::Array(a) => {
            a.iter().enumerate().map(|(i, v)| (i.to_string(), v.clone())).collect()
        }
        _ => vec![],
    }
}

fn rebuild(node: &serde_json::Value, items: &[(String, serde_json::Value)]) -> serde_json::Value {
    match node {
        serde_json::Value::Object(_) => {
            let mut m = serde_json::Map::new();
            for (k, v) in items {
                m.insert(k.clone(), v.clone());
            }
            serde_json::Value::Object(m)
        }
        serde_json::Value::Array(_) => {
            serde_json::Value::Array(items.iter().map(|(_, v)| v.clone()).collect())
        }
        _ => serde_json::Value::Null,
    }
}

File: go/getx.go
package jostraca

import (
	"reflect"
	"regexp"
	"strings"
	"sync"
)

// GetXPath is the narrower variant of GetX taking an explicit
// []string token sequence. Equivalent to GetX(root, []string{...}).
// Use when the path is already split.
func GetXPath(root any, tokens []string) any {
	return GetX(root, tokens)
}

// GetXS is the narrower variant of GetX taking a string path.
// Equivalent to GetX(root, "..."). Use when you want the typed
// signature surfaced to call sites.
func GetXS(root any, path string) any {
	return GetX(root, path)
}

// GetX is the rich-path lookup ported from getx in ts/src/util/basic.ts.
// Supports dot/space-separated navigation, ancestry (`:`), comparison
// filters (`=`, `!=`, `<`, `<=`, `>`, `>=`, `==`, `~`), array filters
// (`?`), array indexing, and quoted segments.
//
// Every step is jsProp: an own property of a map, a slice or a string.
// As in TS, a key that is present with a nil value is not the same as an
// absent one: the walk continues through a present nil, and only an
// absent key is a miss.
//
// Returns nil for any miss or invalid path; otherwise the matched value.
func GetX(root any, path any) any {
	if root == nil {
		return nil
	}
	rv := reflect.ValueOf(root)
	if rv.Kind() != reflect.Map && rv.Kind() != reflect.Slice && rv.Kind() != reflect.Array {
		return nil
	}

	var tokens []string
	switch p := path.(type) {
	case nil:
		return nil
	case string:
		tokens = getxTokenize(p)
	case []string:
		tokens = p
	case []any:
		tokens = make([]string, len(p))
		for i, x := range p {
			tokens[i] = jsString(x)
		}
	default:
		return nil
	}

	out := getxWalk(root, tokens)
	if isUndefined(out) {
		return nil
	}
	return out
}

// getxWalk is the token loop of TS getx. jsUndefined stands for TS's
// undefined wherever the loop tells it apart from null.
func getxWalk(root any, tokens []string) any {
	var node any = root
	var out any = jsUndefined
	ancestry := false

	for i := 0; i < len(tokens) && !isUndefined(node); i++ {
		t0 := tokens[i]
		hasT1 := i+1 < len(tokens)
		var t1 string
		if hasT1 {
			t1 = tokens[i+1]
		}

		if t1 != "" && getxIsCompareOp(t1) {
			val := getxStep(node, t0)
			var arg any = jsUndefined
			if i+2 < len(tokens) {
				arg = getxArg(tokens[i+2])
			}
			if getxCompare(val, t1, arg) {
				i += 2
			} else {
				node = jsUndefined
			}
			if !(ancestry && !isUndefined(node)) {
				out = node
			}
			continue
		}

		if hasT1 && t1 == ":" {
			// A colon followed by `=` is not an ancestry op.
			if !(i+2 < len(tokens) && tokens[i+2] == "=") {
				if !ancestry {
					out = node
				}
				node = getxStep(node, t0)
				if isUndefined(node) {
					out = jsUndefined
				}
			}
			ancestry = true
			i++
			continue
		}

		if t0 == "?" {
			ftokens := tokens[i+1:]
			// Find filter end: two adjacent identifier tokens mark the end.
			j := 0
			for ; j < len(ftokens); j++ {
				if j+1 < len(ftokens) &&
					getxIsIdent(ftokens[j]) && getxIsIdent(ftokens[j+1]) {
					j++
					break
				}
			}
			ftokens = ftokens[:j]

			node = getxFilter(node, ftokens)
			out = node
			i += len(ftokens)
			continue
		}

		if hasT1 {
			node = getxStep(node, t0)
			if ancestry {
				ancestry = false
				if isUndefined(node) {
					out = jsUndefined
				}
				node = out
			}
			continue
		}

		// Last token.
		node = getxStep(node, t0)
		if !(ancestry && !isUndefined(node)) {
			out = node
		}
	}

	return out
}

// getxStep is one step of the walk: jsUndefined for an absent key.
func getxStep(node any, key string) any {
	if v, ok := jsProp(node, key); ok {
		return v
	}
	return jsUndefined
}

// jsSpaceClass is JavaScript's \s, which RE2's ASCII \s is not: it adds
// \v, U+00A0, U+1680, U+2000..U+200A, U+2028, U+2029, U+202F, U+205F,
// U+3000 and U+FEFF.
const jsSpaceClass = `[\t\n\x0B\f\r \x{a0}\x{1680}\x{2000}-\x{200a}\x{2028}\x{2029}\x{202f}\x{205f}\x{3000}\x{feff}]`

// getxTokenRE is GETX_TOKEN_RE from ts/src/util/basic.ts with JavaScript's
// \s spelled out, and JavaScript's '.' (which excludes \r, U+2028 and
// U+2029 as well as \n) in the quoted-string escape.
var getxTokenRE = regexp.MustCompile(jsSpaceClass + `*("(\\[^\n\r\x{2028}\x{2029}]|[^"\\])*"|[\w\d_]+|` +
	jsSpaceClass + `+|[^\w\d_]+)` + jsSpaceClass + `*`)

var jsSpaceRE = regexp.MustCompile(jsSpaceClass)

// getxTokenize splits a string path as TS getx does: a token that
// CONTAINS whitespace or a '.' is dropped, so 'a. b', 'a.?b' and 'a.$'
// all read as 'a b' or 'a', and quotes are stripped only from a token
// matching ^"[^"]+"$.
func getxTokenize(p string) []string {
	if toks, ok := getxTokenizePlain(p); ok {
		return toks
	}
	getxTokenCacheMu.Lock()
	toks, ok := getxTokenCache[p]
	getxTokenCacheMu.Unlock()
	if ok {
		return toks
	}
	toks = getxTokenizeRE(p)
	getxTokenCacheMu.Lock()
	if len(getxTokenCache) >= getxTokenCacheMax {
		getxTokenCache = make(map[string][]string, getxTokenCacheMax)
	}
	getxTokenCache[p] = toks
	getxTokenCacheMu.Unlock()
	return toks
}

// Tokens are never mutated by the walk, so a cached slice can be shared.
const getxTokenCacheMax = 1000

var (
	getxTokenCacheMu sync.Mutex
	getxTokenCache   = make(map[string][]string, getxTokenCacheMax)
)

// getxTokenizePlain is the regex's answer for the common path made only of
// ASCII word characters, '.' and ' ': every run of word characters is a
// token, and every run of the other two is a dropped separator.
func getxTokenizePlain(p string) ([]string, bool) {
	out := []string{}
	start := -1
	for i := 0; i < len(p); i++ {
		c := p[i]
		switch {
		case (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || (c >= '0' && c <= '9') || c == '_':
			if start < 0 {
				start = i
			}
		case c == '.' || c == ' ':
			if start >= 0 {
				out = append(out, p[start:i])
				start = -1
			}
		default:
			return nil, false
		}
	}
	if start >= 0 {
		out = append(out, p[start:])
	}
	return out, true
}

func getxTokenizeRE(p string) []string {
	out := []string{}
	for _, m := range getxTokenRE.FindAllStringSubmatch(p, -1) {
		tok := m[1]
		if strings.Contains(tok, ".") || jsSpaceRE.MatchString(tok) {
			continue
		}
		out = append(out, jsUnquote(tok))
	}
	return out
}

// jsUnquote strips the quotes from a token matching ^"[^"]+"$, and returns
// anything else unchanged.
func jsUnquote(tok string) string {
	if len(tok) >= 3 && tok[0] == '"' && tok[len(tok)-1] == '"' &&
		!strings.Contains(tok[1:len(tok)-1], `"`) {
		return tok[1 : len(tok)-1]
	}
	return tok
}

func getxIsCompareOp(t string) bool {
	switch t {
	case "=", "!=", "<", "<=", ">", ">=", "==", "~":
		return true
	}
	return false
}

// getxIsIdent is TS's filter-end test, /[\w\d_]+/ unanchored: the token
// CONTAINS an ASCII word character, so 't-w' counts.
func getxIsIdent(t string) bool {
	for i := 0; i < len(t); i++ {
		c := t[i]
		if (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') ||
			(c >= '0' && c <= '9') || c == '_' {
			return true
		}
	}
	return false
}

// getxArg coerces a comparison argument as TS does: only 'true' and
// 'false' become booleans, and quotes come off ^"[^"]+"$. Nothing is
// parsed as a number -- 'NaN', 'inf' and '1e1' stay strings.
func getxArg(tok string) any {
	switch tok {
	case "true":
		return true
	case "false":
		return false
	}
	return jsUnquote(tok)
}

// getxCompare applies op with JavaScript's semantics: '=' is ==, '==' is
// ===, '!=' is !=, the orderings are JS relational comparison (strings by
// UTF-16 code unit, anything else by ToNumber), and '~' is
// String(val).match(RegExp(arg)). val and arg may be jsUndefined.
//
// '~' compiles with RE2, so a pattern RE2 rejects is a non-match where JS
// would throw, and the two regex dialects differ at their edges.
func getxCompare(val any, op string, arg any) bool {
	switch op {
	case "<":
		lt, undef := jsLessThan(val, arg)
		return !undef && lt
	case ">":
		lt, undef := jsLessThan(arg, val)
		return !undef && lt
	case "<=":
		lt, undef := jsLessThan(arg, val)
		return !undef && !lt
	case ">=":
		lt, undef := jsLessThan(val, arg)
		return !undef && !lt
	case "=":
		return jsLooseEqual(val, arg)
	case "==":
		return jsStrictEqual(val, arg)
	case "!=":
		return !jsLooseEqual(val, arg)
	case "~":
		pattern := "(?:)"
		if !isUndefined(arg) {
			pattern = jsString(arg)
		}
		re, err := regexp.Compile(pattern)
		if err != nil {
			return false
		}
		return re.MatchString(jsString(val))
	}
	return false
}

// getxFilter keeps the RAW children of a map or slice for which the
// filter path resolves to something other than nil, as TS getx '?' does:
// a slice gives a []any in order, a map a map[string]any. Filtering
// anything else is a miss.
func getxFilter(node any, ftokens []string) any {
	switch v := node.(type) {
	case map[string]any:
		out := map[string]any{}
		for k, c := range v {
			if GetX(c, ftokens) != nil {
				out[k] = c
			}
		}
		return out
	case []any:
		out := make([]any, 0, len(v))
		for _, c := range v {
			if GetX(c, ftokens) != nil {
				out = append(out, c)
			}
		}
		return out
	}
	rv := reflect.ValueOf(node)
	switch rv.Kind() {
	case reflect.Map:
		out := map[string]any{}
		iter := rv.MapRange()
		for iter.Next() {
			c := iter.Value().Interface()
			if GetX(c, ftokens) != nil {
				out[jsJSONMapKey(iter.Key())] = c
			}
		}
		return out
	case reflect.Slice, reflect.Array:
		out := make([]any, 0, rv.Len())
		for i := 0; i < rv.Len(); i++ {
			c := rv.Index(i).Interface()
			if GetX(c, ftokens) != nil {
				out = append(out, c)
			}
		}
		return out
	}
	return jsUndefined
}
